src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs

```diff
@@ -618,445 +501,1122 @@ private static PackOptions GetPackOptions(PackageSpec packageSpec, JObject rawPa
-                Dependencies = new List<LibraryDependency>(),
-                Imports = importFrameworks,
-                Warn = GetWarnSetting(properties),
-                AssetTargetFallback = assetTargetFallback,
-                RuntimeIdentifierGraphPath = GetRuntimeIdentifierGraphPath(properties)
-            };
-
-            PopulateDependencies(
-                packageSpec.FilePath,
-                targetFrameworkInformation.Dependencies,
-                properties,
-                "dependencies",
-                isGacOrFrameworkReference: false);
-
-            PopulateDownloadDependencies(
-                targetFrameworkInformation.DownloadDependencies,
-                properties,
-                packageSpec.FilePath);
-
-            PopulateCentralDependencies(
-               packageSpec.FilePath,
-               targetFrameworkInformation,
-               properties);
-
-            var frameworkAssemblies = new List<LibraryDependency>();
-            PopulateDependencies(
-                packageSpec.FilePath,
-                frameworkAssemblies,
-                properties,
-                "frameworkAssemblies",
-                isGacOrFrameworkReference: true);
-
-            PopulateFrameworkReferences(
-                targetFrameworkInformation.FrameworkReferences,
-                properties,
-                "frameworkReferences",
-                packageSpec.FilePath);
-
-            frameworkAssemblies.ForEach(d => targetFrameworkInformation.Dependencies.Add(d));
+                msbuildMetadata.ConfigFilePaths = configFilePaths;
+            }
 
-            packageSpec.TargetFrameworks.Add(targetFrameworkInformation);
+            msbuildMetadata.CrossTargeting = crossTargeting;
+
+            if (fallbackFolders != null)
+            {
+                msbuildMetadata.FallbackFolders = fallbackFolders;
+            }
+
+            if (files != null)
+            {
+                msbuildMetadata.Files = files;
+            }
+
+            msbuildMetadata.LegacyPackagesDirectory = legacyPackagesDirectory;
+
+            if (originalTargetFrameworks != null)
+            {
+                msbuildMetadata.OriginalTargetFrameworks = originalTargetFrameworks;
+            }
+
+            msbuildMetadata.OutputPath = outputPath;
+            msbuildMetadata.PackagesPath = packagesPath;
+            msbuildMetadata.ProjectJsonPath = projectJsonPath;
+            msbuildMetadata.ProjectName = projectName;
+            msbuildMetadata.ProjectPath = projectPath;
+
+            if (projectStyle.HasValue)
+            {
+                msbuildMetadata.ProjectStyle = projectStyle.Value;
+            }
+
+            msbuildMetadata.ProjectUniqueName = projectUniqueName;
+
+            if (restoreLockProperties != null)
+            {
+                msbuildMetadata.RestoreLockProperties = restoreLockProperties;
+            }
+
+            msbuildMetadata.SkipContentFileWrite = skipContentFileWrite;
```

Comment:
Seems like having the parsing and the setting of the data seperate, is somewhat risky... what if you harvest data, but forget to set it.  I would lean towards setting the value while harvesting, when possible...and having a minimal amount of state set after the harvesting operations are all done.

Reply:
For several of the places you pointed this out elsewhere, I've fixed them.  However, for restore metadata this unavoidable:  we have to read the project style before we can even create `msbuildMetadata`.  It could be either an instance of `PackagesConfigProjectRestoreMetadata` or `ProjectRestoreMetadata` --- we don't know until we've read all properties.